proxy/envoy/discovery.go

```diff
@@ -87,27 +109,99 @@ func (ds *DiscoveryService) Run() {
 
 // ListEndpoints responds to SDS requests
 func (ds *DiscoveryService) ListEndpoints(request *restful.Request, response *restful.Response) {
-	key := request.PathParameter("service-key")
-	hostname, ports, tags := model.ParseServiceKey(key)
-	out := make([]host, 0)
+	hostname, ports, tags := model.ParseServiceKey(request.PathParameter(ServiceKey))
+	// envoy expects an empty array if no hosts are available
+	out := make([]*host, 0)
 	for _, ep := range ds.services.Instances(hostname, ports.GetNames(), tags) {
-		out = append(out, host{
+		out = append(out, &host{
 			Address: ep.Endpoint.Address,
 			Port:    ep.Endpoint.Port,
 		})
 	}
-	if err := response.WriteEntity(hosts{out}); err != nil {
+	if err := response.WriteEntity(hosts{Hosts: out}); err != nil {
 		glog.Warning(err)
 	}
 }
 
-// ListClusters responds to CDS requests
+// ListClusters responds to CDS requests for outbound HTTP clusters
 func (ds *DiscoveryService) ListClusters(request *restful.Request, response *restful.Response) {
-	_ = ds.services.Services()
-	// TODO: fix this
-	/*
-		if err := response.WriteEntity(clusters{buildClusters(svc)}); err != nil {
-			glog.Warning(err)
-		}
-	*/
+	if serviceCluster := request.PathParameter(ServiceCluster); serviceCluster != IstioServiceCluster {
+		errorResponse(fmt.Sprintf("Unexpected %s %q", ServiceCluster, serviceCluster), response)
+		return
+	}
+
+	// service-node holds the IP address
+	ip := request.PathParameter(ServiceNode)
+
+	// CDS computes clusters that are referenced by RDS routes for a particular proxy node
+	// TODO: this implementation is inefficient as it is recomputing all the routes for all proxies
+	// There is a lot of potential to cache and reuse cluster definitions across proxies and also
+	// skip computing the actual HTTP routes
+	httpRouteConfigs, _ := buildRoutes(&ProxyContext{
+		Discovery:  ds.services,
+		Config:     ds.config,
+		MeshConfig: ds.mesh,
+		Addrs:      map[string]bool{ip: true},
+	})
+
+	clusters := make(Clusters, 0)
+	for _, httpRouteConfig := range httpRouteConfigs {
+		clusters = append(clusters, httpRouteConfig.filterClusters(func(cluster *Cluster) bool {
+			return cluster.outbound
+		})...)
+	}
+
+	// de-duplicate and canonicalize clusters
+	clusters = clusters.Normalize()
+
+	for _, cluster := range clusters {
+		insertDestinationPolicy(ds.config, cluster)
+	}
+
+	if err := response.WriteEntity(ClusterManager{Clusters: clusters}); err != nil {
+		glog.Warning(err)
+	}
+}
+
+// ListRoutes responds to RDS requests, used by HTTP routes
+func (ds *DiscoveryService) ListRoutes(request *restful.Request, response *restful.Response) {
+	if serviceCluster := request.PathParameter(ServiceCluster); serviceCluster != IstioServiceCluster {
```

Comment:
This is effectively disabling all source based routing isn't it? Why can't we run the RDS and CDS on localhost next to Envoy itself so that the proxy agents know their parent services? This is where the explicit service affinity declaration comes in.